{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1c8c642-b9c3-45f5-9764-7c8a0532527c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if the files are on Anvil / Terra workspace \n",
    "# please run it from the workspace analysis to avoid bucket access permission issues\n",
    "# or use the gcloud login auth interface to authenticate yourself"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a8633cc-1724-42a1-ad4b-cb4b77856714",
   "metadata": {},
   "source": [
    "installations:\n",
    "\n",
    "1. pip install synapseclient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e395f4-c113-4e1a-b5d1-e97ed59417d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "local_metadata_tsv = \"proc.tsv\"\n",
    "local_clusters_fld = os.path.join(os.getcwd(),\"clusters\")\n",
    "os.makedirs(local_clusters_fld, exist_ok=True)\n",
    "local_cell_type_annotation_fld = os.path.join(os.getcwd(),\"labels\")\n",
    "os.makedirs(local_cell_type_annotation_fld, exist_ok=True)\n",
    "syn_encode_cell_types_annotation_dataset_id = 'syn34271785' #ENCODE DATASET OF THE ANNOTATION\n",
    "local_path_to_download = os.path.join(os.getcwd(),\"fragment_files\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0e0ad4c-71d1-42ce-81ed-12aa748c7710",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_file(filename):\n",
    "  print(\"remove_file method: {}\".format(filename))\n",
    "  if os.path.exists(filename):\n",
    "    os.remove(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d86f17bc-2d0e-4a59-a60e-46aaeba3a151",
   "metadata": {},
   "outputs": [],
   "source": [
    "%run synapse_utils_from_analysis.ipynb\n",
    "%run tsv_files_utils_from_analysis.ipynb "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ddd48e-496f-48f0-b0b4-15a83f82b6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the proc.tsv object. Validation that the annoation file is available\n",
    "local_metadata_file_location = os.path.join(os.getcwd(),local_metadata_tsv)\n",
    "if os.path.exists(local_metadata_file_location) == False:\n",
    "    print(\"please copy your proc.tsv locally {}\".format(local_metadata_file_location))\n",
    "else:\n",
    "    print(\"proc.tsv is available locally {}\".format(local_metadata_file_location))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63e119a7-cfc1-4d7f-82c0-41d0a2c4af7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace this with your own data annoation\n",
    "# TODO: create a method that reads from other annoation files\n",
    "df_encode_cell_types = read_proc_tsv(local_metadata_file_location)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ae77f2-8da9-49f6-9b88-1ec6f98674bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_encode_cell_types.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc8079af-fe98-42dd-a28b-8b9911a44210",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_encode_cell_types['atac_dataset'].to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1517927-98be-408a-96c7-ebb3d34b8b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the list of the fragment\n",
    "local_fragment_files_path = os.path.join(os.getcwd(),\"fragment_files\",\"*\",\"*\")\n",
    "print(\"local_fragment_files_path\")\n",
    "local_fragment_files = !ls $local_fragment_files_path\n",
    "local_fragment_files = list(local_fragment_files)\n",
    "local_fragment_files[4]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d47b3b1d-2823-46e5-ab05-15a7fa157a41",
   "metadata": {},
   "source": [
    "- check that the files in the fragment files has matching synapse data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "584236eb-6405-4c0d-998d-cb7c6727604c",
   "metadata": {},
   "outputs": [],
   "source": [
    "local_files_atac_dataset = [x.split(\"_\")[-1].split(\".\")[0] for x in local_fragment_files]\n",
    "local_files_atac_dataset[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac47fefe-cc12-4165-8409-4a0895d0dd0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_encode_cell_types_atac_dataset = df_encode_cell_types[\"atac_dataset\"].tolist()\n",
    "df_encode_cell_types_atac_dataset[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "042a0d23-f26e-4085-acc9-c9477b648c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# atac dataset in both\n",
    "atac_dataset_to_fragment = list(set(local_files_atac_dataset) & set(df_encode_cell_types_atac_dataset))\n",
    "atac_dataset_to_fragment[0:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18ad0b98-7851-439a-9890-def601acb23e",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(atac_dataset_to_fragment)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71509b0a-cf2d-4fdd-854b-45c4e3187617",
   "metadata": {},
   "outputs": [],
   "source": [
    "local_fragment_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e3bd2c0-a7b9-404e-877b-113ced19d7de",
   "metadata": {},
   "outputs": [],
   "source": [
    "local_fragment_files_path_match =[x for x in local_fragment_files if x.split(\"_\")[-1].split(\".\")[0] in atac_dataset_to_fragment]\n",
    "local_fragment_files_path_match[0:3]\n",
    "len(local_fragment_files)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c97c222f-19c8-4feb-b596-65dea63d5762",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this will check for which fragments do we have annotation data\n",
    "df_encode_atac_dataset_relevant_rows = df_encode_cell_types[df_encode_cell_types['atac_dataset'].isin(atac_dataset_to_fragment)]\n",
    "df_encode_atac_dataset_relevant_rows.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "450884e4-a3a1-49c2-ad67-995fb0483bf3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished clustering local_fragment_file /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119922_ENCSR411MLR.atac.filter.fragments.hg38.tsv.gz/ENCSR411MLR.atac.filter.fragments.hg38.tsv.gz by cell type. for types ['T_lymphocyte', 'Macrophage', 'Vascular_smooth_muscle_cell.1', 'Vascular_smooth_muscle_cell.2', 'Endothelial_cell']\n",
      "total missing bc are 20174691\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119924_ENCSR052DKH.atac.filter.fragments.hg38.tsv.gz/ENCSR052DKH.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR052DKH\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000043\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46650139 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46650145 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471000>\n",
      "syn_file_id_for_atac_dataset_id is syn46650146\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR052DKH_syn46650146\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR052DKH_syn46650146/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR052DKH_syn46650146/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR052DKH/tagAlign_ENCSR052DKH_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR052DKH/tagAlign_ENCSR052DKH_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR052DKH/tagAlign_ENCSR052DKH_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR052DKH/tagAlign_ENCSR052DKH_Tuft_cell.tsv', 'Tuft_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119927_ENCSR845QFX.atac.filter.fragments.hg38.tsv.gz/ENCSR845QFX.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR845QFX\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000020\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46646366 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46646398 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46646416\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR845QFX_syn46646416\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR845QFX_syn46646416/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR845QFX_syn46646416/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 6\n",
      "tagAlign_exists is [True, True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR845QFX/tagAlign_ENCSR845QFX_Luminal_epithelial_cell.3.tsv', 'Luminal_epithelial_cell.3'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR845QFX/tagAlign_ENCSR845QFX_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR845QFX/tagAlign_ENCSR845QFX_Luminal_epithelial_cell.1.tsv', 'Luminal_epithelial_cell.1'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR845QFX/tagAlign_ENCSR845QFX_Luminal_epithelial_cell.2.tsv', 'Luminal_epithelial_cell.2'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR845QFX/tagAlign_ENCSR845QFX_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR845QFX/tagAlign_ENCSR845QFX_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119934_ENCSR660NEE.atac.filter.fragments.hg38.tsv.gz/ENCSR660NEE.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR660NEE\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000023\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46647276 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46647284 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470ca0>\n",
      "syn_file_id_for_atac_dataset_id is syn46647285\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR660NEE_syn46647285\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR660NEE_syn46647285/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR660NEE_syn46647285/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 8\n",
      "tagAlign_exists is [True, True, True, True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Smooth_muscle_cell.tsv', 'Smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Schwann_cell.tsv', 'Schwann_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Mast_cell.tsv', 'Mast_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR660NEE/tagAlign_ENCSR660NEE_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119935_ENCSR023FME.atac.filter.fragments.hg38.tsv.gz/ENCSR023FME.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR023FME\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000011\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46645395 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46645400 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471000>\n",
      "syn_file_id_for_atac_dataset_id is syn46645403\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR023FME_syn46645403\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR023FME_syn46645403/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR023FME_syn46645403/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR023FME/tagAlign_ENCSR023FME_Endothelial_cell.tsv', 'Endothelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR023FME/tagAlign_ENCSR023FME_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR023FME/tagAlign_ENCSR023FME_Type_I_skeletal_myocyte.tsv', 'Type_I_skeletal_myocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR023FME/tagAlign_ENCSR023FME_Basal_epithelial_cell.tsv', 'Basal_epithelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119936_ENCSR453TVZ.atac.filter.fragments.hg38.tsv.gz/ENCSR453TVZ.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR453TVZ\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000027\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46647506 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46647511 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46647517\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR453TVZ_syn46647517\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR453TVZ_syn46647517/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR453TVZ_syn46647517/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR453TVZ/tagAlign_ENCSR453TVZ_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR453TVZ/tagAlign_ENCSR453TVZ_Basal_epithelial_cell.1.tsv', 'Basal_epithelial_cell.1'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR453TVZ/tagAlign_ENCSR453TVZ_Basal_epithelial_cell.3.tsv', 'Basal_epithelial_cell.3'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR453TVZ/tagAlign_ENCSR453TVZ_Basal_epithelial_cell.2.tsv', 'Basal_epithelial_cell.2')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119939_ENCSR868ZEI.atac.filter.fragments.hg38.tsv.gz/ENCSR868ZEI.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR868ZEI\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000029\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46647621 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46647632 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46647634\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR868ZEI_syn46647634\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR868ZEI_syn46647634/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR868ZEI_syn46647634/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 9\n",
      "tagAlign_exists is [True, True, True, True, True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Smooth_muscle_cell.tsv', 'Smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Myofibroblast.tsv', 'Myofibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Schwann_cell.tsv', 'Schwann_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Mast_cell.tsv', 'Mast_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR868ZEI/tagAlign_ENCSR868ZEI_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119946_ENCSR474TGL.atac.filter.fragments.hg38.tsv.gz/ENCSR474TGL.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR474TGL\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000033\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46647847 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46647854 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471000>\n",
      "syn_file_id_for_atac_dataset_id is syn46647859\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR474TGL_syn46647859\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR474TGL_syn46647859/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR474TGL_syn46647859/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR474TGL/tagAlign_ENCSR474TGL_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR474TGL/tagAlign_ENCSR474TGL_Endothelial_cell.tsv', 'Endothelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR474TGL/tagAlign_ENCSR474TGL_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR474TGL/tagAlign_ENCSR474TGL_Basal_epithelial_cell.tsv', 'Basal_epithelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119947_ENCSR538FPK.atac.filter.fragments.hg38.tsv.gz/ENCSR538FPK.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR538FPK\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000031\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46647742 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46647744 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471000>\n",
      "syn_file_id_for_atac_dataset_id is syn46647748\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR538FPK_syn46647748\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR538FPK_syn46647748/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR538FPK_syn46647748/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 6\n",
      "tagAlign_exists is [True, True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR538FPK/tagAlign_ENCSR538FPK_Smooth_muscle_cell.tsv', 'Smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR538FPK/tagAlign_ENCSR538FPK_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR538FPK/tagAlign_ENCSR538FPK_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR538FPK/tagAlign_ENCSR538FPK_Cardiomyocyte.tsv', 'Cardiomyocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR538FPK/tagAlign_ENCSR538FPK_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR538FPK/tagAlign_ENCSR538FPK_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119948_ENCSR181XXQ.atac.filter.fragments.hg38.tsv.gz/ENCSR181XXQ.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR181XXQ\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000039\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46648345 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46648351 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46648356\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR181XXQ_syn46648356\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR181XXQ_syn46648356/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR181XXQ_syn46648356/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR181XXQ/tagAlign_ENCSR181XXQ_Adipocyte.tsv', 'Adipocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR181XXQ/tagAlign_ENCSR181XXQ_Endothelial_cell.tsv', 'Endothelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR181XXQ/tagAlign_ENCSR181XXQ_Mesothelial_cell.tsv', 'Mesothelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR181XXQ/tagAlign_ENCSR181XXQ_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119951_ENCSR776EOI.atac.filter.fragments.hg38.tsv.gz/ENCSR776EOI.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR776EOI\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000045\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46650260 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46650266 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46650268\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR776EOI_syn46650268\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR776EOI_syn46650268/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR776EOI_syn46650268/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR776EOI/tagAlign_ENCSR776EOI_Cardiomyocyte.tsv', 'Cardiomyocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR776EOI/tagAlign_ENCSR776EOI_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR776EOI/tagAlign_ENCSR776EOI_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR776EOI/tagAlign_ENCSR776EOI_Mesothelial_cell.tsv', 'Mesothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119956_ENCSR848EID.atac.filter.fragments.hg38.tsv.gz/ENCSR848EID.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR848EID\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000052\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46650828 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46650840 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471120>\n",
      "syn_file_id_for_atac_dataset_id is syn46650847\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR848EID_syn46650847\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR848EID_syn46650847/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR848EID_syn46650847/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 5\n",
      "tagAlign_exists is [True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR848EID/tagAlign_ENCSR848EID_Luminal_epithelial_cell.tsv', 'Luminal_epithelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR848EID/tagAlign_ENCSR848EID_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR848EID/tagAlign_ENCSR848EID_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR848EID/tagAlign_ENCSR848EID_Gastric_chief_cell.tsv', 'Gastric_chief_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR848EID/tagAlign_ENCSR848EID_Fibroblast.tsv', 'Fibroblast')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119958_ENCSR367GKP.atac.filter.fragments.hg38.tsv.gz/ENCSR367GKP.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR367GKP\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000047\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46650437 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46650452 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470280>\n",
      "syn_file_id_for_atac_dataset_id is syn46650459\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR367GKP_syn46650459\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR367GKP_syn46650459/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR367GKP_syn46650459/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 5\n",
      "tagAlign_exists is [True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR367GKP/tagAlign_ENCSR367GKP_Smooth_muscle_cell.tsv', 'Smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR367GKP/tagAlign_ENCSR367GKP_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR367GKP/tagAlign_ENCSR367GKP_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR367GKP/tagAlign_ENCSR367GKP_Schwann_cell.tsv', 'Schwann_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR367GKP/tagAlign_ENCSR367GKP_Fibroblast.tsv', 'Fibroblast')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119960_ENCSR397LYX.atac.filter.fragments.hg38.tsv.gz/ENCSR397LYX.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR397LYX\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000053\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46650901 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46650905 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470ca0>\n",
      "syn_file_id_for_atac_dataset_id is syn46650909\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR397LYX_syn46650909\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR397LYX_syn46650909/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR397LYX_syn46650909/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 3\n",
      "tagAlign_exists is [True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR397LYX/tagAlign_ENCSR397LYX_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR397LYX/tagAlign_ENCSR397LYX_Endothelial_cell.tsv', 'Endothelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR397LYX/tagAlign_ENCSR397LYX_Basal_epithelial_cell.tsv', 'Basal_epithelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119967_ENCSR796RXX.atac.filter.fragments.hg38.tsv.gz/ENCSR796RXX.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR796RXX\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000057\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46651207 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46651217 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471120>\n",
      "syn_file_id_for_atac_dataset_id is syn46651221\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR796RXX_syn46651221\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR796RXX_syn46651221/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR796RXX_syn46651221/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 3\n",
      "tagAlign_exists is [True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR796RXX/tagAlign_ENCSR796RXX_Follicular_cell.tsv', 'Follicular_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR796RXX/tagAlign_ENCSR796RXX_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR796RXX/tagAlign_ENCSR796RXX_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119968_ENCSR302CMX.atac.filter.fragments.hg38.tsv.gz/ENCSR302CMX.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR302CMX\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000061\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46651679 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46651687 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46651692\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR302CMX_syn46651692\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR302CMX_syn46651692/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR302CMX_syn46651692/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR302CMX/tagAlign_ENCSR302CMX_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR302CMX/tagAlign_ENCSR302CMX_Smooth_muscle_cell.tsv', 'Smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR302CMX/tagAlign_ENCSR302CMX_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR302CMX/tagAlign_ENCSR302CMX_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119970_ENCSR822GXQ.atac.filter.fragments.hg38.tsv.gz/ENCSR822GXQ.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR822GXQ\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000055\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46651019 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46651029 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470ca0>\n",
      "syn_file_id_for_atac_dataset_id is syn46651032\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR822GXQ_syn46651032\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR822GXQ_syn46651032/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR822GXQ_syn46651032/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 6\n",
      "tagAlign_exists is [False, False, False, False, False, False]\n",
      "!!!output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR822GXQ/tagAlign_ENCSR822GXQ_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR822GXQ/tagAlign_ENCSR822GXQ_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR822GXQ/tagAlign_ENCSR822GXQ_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR822GXQ/tagAlign_ENCSR822GXQ_Basal_epithelial_cell.tsv', 'Basal_epithelial_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR822GXQ/tagAlign_ENCSR822GXQ_Fibroblast.tsv', 'Fibroblast'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR822GXQ/tagAlign_ENCSR822GXQ_Endothelial_cell.tsv', 'Endothelial_cell')]. open files\n",
      "open local_fragment_file /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119970_ENCSR822GXQ.atac.filter.fragments.hg38.tsv.gz/ENCSR822GXQ.atac.filter.fragments.hg38.tsv.gz\n",
      "finished clustering local_fragment_file /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119970_ENCSR822GXQ.atac.filter.fragments.hg38.tsv.gz/ENCSR822GXQ.atac.filter.fragments.hg38.tsv.gz by cell type. for types ['T_lymphocyte', 'Vascular_smooth_muscle_cell', 'Macrophage', 'Basal_epithelial_cell', 'Fibroblast', 'Endothelial_cell']\n",
      "total missing bc are 66964161\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119971_ENCSR205TUH.atac.filter.fragments.hg38.tsv.gz/ENCSR205TUH.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR205TUH\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000063\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46651886 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46651919 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470ca0>\n",
      "syn_file_id_for_atac_dataset_id is syn46651930\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR205TUH_syn46651930\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR205TUH_syn46651930/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR205TUH_syn46651930/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 2\n",
      "tagAlign_exists is [True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR205TUH/tagAlign_ENCSR205TUH_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR205TUH/tagAlign_ENCSR205TUH_Schwann_cell.tsv', 'Schwann_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119982_ENCSR718WGE.atac.filter.fragments.hg38.tsv.gz/ENCSR718WGE.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR718WGE\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000070\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46653802 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46653815 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e471000>\n",
      "syn_file_id_for_atac_dataset_id is syn46653820\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR718WGE_syn46653820\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR718WGE_syn46653820/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR718WGE_syn46653820/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 4\n",
      "tagAlign_exists is [True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR718WGE/tagAlign_ENCSR718WGE_Type_II_pneumocyte.tsv', 'Type_II_pneumocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR718WGE/tagAlign_ENCSR718WGE_Macrophage.tsv', 'Macrophage'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR718WGE/tagAlign_ENCSR718WGE_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR718WGE/tagAlign_ENCSR718WGE_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119983_ENCSR506YMX.atac.filter.fragments.hg38.tsv.gz/ENCSR506YMX.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR506YMX\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000065\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46652361 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46652394 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470a60>\n",
      "syn_file_id_for_atac_dataset_id is syn46652413\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR506YMX_syn46652413\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR506YMX_syn46652413/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR506YMX_syn46652413/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 5\n",
      "tagAlign_exists is [True, True, True, True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR506YMX/tagAlign_ENCSR506YMX_T_lymphocyte.tsv', 'T_lymphocyte'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR506YMX/tagAlign_ENCSR506YMX_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR506YMX/tagAlign_ENCSR506YMX_Tuft_cell.tsv', 'Tuft_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR506YMX/tagAlign_ENCSR506YMX_Goblet_cell.tsv', 'Goblet_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR506YMX/tagAlign_ENCSR506YMX_Luminal_epithelial_cell.tsv', 'Luminal_epithelial_cell')] is at work or was already downloaded. continue\n",
      "!!!!!local_fragment_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/fragment_files/syn52118179_syn52119985_ENCSR455CVZ.atac.filter.fragments.hg38.tsv.gz/ENCSR455CVZ.atac.filter.fragments.hg38.tsv.gz\n",
      "file_atac_dataset_id is ENCSR455CVZ\n",
      "syn_folder_name_for_cell_types in synpase is EAID_000074\n",
      "get_file_names parameters are: folder_id syn34271785 and current_depth 0\n",
      "get_file_names parameters are: folder_id syn46654105 and current_depth 1\n",
      "get_file_names parameters are: folder_id syn46654112 and current_depth 2\n",
      "folder_entities are <generator object Synapse.getChildren at 0x29e470ca0>\n",
      "syn_file_id_for_atac_dataset_id is syn46654113\n",
      "cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR455CVZ_syn46654113\n",
      "this was already downloaded from synapse\n",
      "full_cell_types_annotation_file_path is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR455CVZ_syn46654113/cell_types.tsv.gz\n",
      "tsv_gz_file is /Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/labels/syn34271785_ENCSR455CVZ_syn46654113/cell_types.tsv.gz\n",
      "number of cell_type_id_names_for_atac_dataset is 2\n",
      "tagAlign_exists is [True, True]\n",
      "output_tagAlign_files_with_names [('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR455CVZ/tagAlign_ENCSR455CVZ_Vascular_smooth_muscle_cell.tsv', 'Vascular_smooth_muscle_cell'), ('/Users/eilaarich-landkof-stanford/Documents/Code/git_kundaje_genomics/DNA/ATAC/single-cell/fragments-file-manipulation/synapse_in_out/clusters/ENCSR455CVZ/tagAlign_ENCSR455CVZ_Endothelial_cell.tsv', 'Endothelial_cell')] is at work or was already downloaded. continue\n"
     ]
    }
   ],
   "source": [
    "%run tsv_files_utils_from_analysis.ipynb\n",
    "%run synapse_utils_from_analysis.ipynb\n",
    "\n",
    "# TODO: for others. will the data come from synpase as well?\n",
    "#local_clusters_fld\n",
    "for local_fragment_file in local_fragment_files:\n",
    "\n",
    "    # Dictionary to store the output file handles with names\n",
    "    output_handles = {}\n",
    "    print(\"!!!!!local_fragment_file is {}\".format(local_fragment_file))\n",
    "    file_atac_dataset_id = local_fragment_file.split(\"_\")[-1].split(\".\")[0]\n",
    "    print(\"file_atac_dataset_id is {}\".format(file_atac_dataset_id))\n",
    "\n",
    "    # Retrieve the synpase ID from the proc.tsv dataframe\n",
    "    syn_folder_name_for_cell_types = df_encode_atac_dataset_relevant_rows[df_encode_atac_dataset_relevant_rows['atac_dataset'] == file_atac_dataset_id]['file'].to_list()[0]\n",
    "    print(\"syn_folder_name_for_cell_types in synpase is {}\".format(syn_folder_name_for_cell_types))\n",
    "    # (starting_folder_id, first_subfolder_name, second_subfolder_name):\n",
    "    # retrieve the location on synpase\n",
    "    syn_file_id_list = get_file_names_in_second_subfolder(syn_encode_cell_types_annotation_dataset_id, \\\n",
    "                                                          syn_folder_name_for_cell_types, \\\n",
    "                                                          \"labels\")\n",
    "    syn_file_id_for_atac_dataset_id = syn_file_id_list[0] # the label's file synID\n",
    "    print('syn_file_id_for_atac_dataset_id is {}'.format(syn_file_id_for_atac_dataset_id))\n",
    "    \n",
    "    # Download the synpase label file:\n",
    "    # Obtain a pointer and download the data to full_labels_file_path\n",
    "    # {}_{}_{}\".format(dataset_id,file_id,file_name))\n",
    "    cell_types_annotation_file_path = os.path.join(local_cell_type_annotation_fld,\"{}_{}_{}\".format(syn_encode_cell_types_annotation_dataset_id, \\\n",
    "                                                                                                    file_atac_dataset_id, \\\n",
    "                                                                                                    syn_file_id_for_atac_dataset_id))\n",
    "\n",
    "    print(\"cell_types_annotation_file_path is {}\".format(cell_types_annotation_file_path))\n",
    "    os_listdir = []\n",
    "    if os.path.exists(cell_types_annotation_file_path):\n",
    "        os.listdir(cell_types_annotation_file_path)\n",
    "    # os_listdir = os.listdir(cell_types_annotation_file_path)\n",
    "    if len(os_listdir) > 0:\n",
    "        print(\"this was already downloaded from synapse\")\n",
    "        full_cell_types_annotation_file_path = os.path.join(cell_types_annotation_file_path,os_listdir[0])\n",
    "    else:\n",
    "        cell_types_annotation_file_synHandle = syn.get(entity=syn_file_id_for_atac_dataset_id,downloadLocation=cell_types_annotation_file_path) \n",
    "        print('cell_types_annotation_file_synHandle is {}'.format(cell_types_annotation_file_synHandle))\n",
    "        full_cell_types_annotation_file_path = cell_types_annotation_file_synHandle.path\n",
    "    \n",
    "    print('full_cell_types_annotation_file_path is {}'.format(full_cell_types_annotation_file_path))\n",
    "    \n",
    "#   # Load the labels data to dataframe. has all the barcodes and the cell types relevant for them\n",
    "    df_cell_types_for_atac_dataset = read_tsv_gz_to_dataframe_skipping_comments_and_empty_lines(full_cell_types_annotation_file_path)\n",
    "    cell_type_id_names_for_atac_dataset = list(set(df_cell_types_for_atac_dataset['cell_type_id'].to_list()))\n",
    "    print(\"number of cell_type_id_names_for_atac_dataset is {}\".format(len(cell_type_id_names_for_atac_dataset)))\n",
    "    # List of output text files with corresponding names\n",
    "    # [(\"output1.txt\", \"file_1\"), (\"output2.txt\", \"file_2\"), (\"output3.txt\", \"file_3\")]\n",
    "    output_tagAlign_files_with_names = [(os.path.join(local_clusters_fld,file_atac_dataset_id,\"tagAlign_{}_{}.tsv\".format(file_atac_dataset_id,cell_type_name_id)),\n",
    "                                        cell_type_name_id) for cell_type_name_id in cell_type_id_names_for_atac_dataset]\n",
    "\n",
    "    tagAlign_exists = [os.path.exists(output_tagAlign_file[0]) for output_tagAlign_file in output_tagAlign_files_with_names]\n",
    "    print(\"tagAlign_exists is {}\".format(tagAlign_exists))\n",
    "    if sum(tagAlign_exists) >0:\n",
    "        print(\"output_tagAlign_files_with_names {} is at work or was already downloaded. continue\".format(output_tagAlign_files_with_names))\n",
    "        continue # either started by annother process or already was processed\n",
    "    else:\n",
    "        print(\"!!!output_tagAlign_files_with_names {}. open files\".format(output_tagAlign_files_with_names))\n",
    "        for tag_file_path, tag_file_cell_type_name in output_tagAlign_files_with_names:\n",
    "            # print(\"tag_file_path is {}\".format(tag_file_path))\n",
    "            # print(\"os.path.dirname(tag_file_path) is {}\".format(os.path.dirname(tag_file_path)))\n",
    "            os.makedirs(os.path.dirname(tag_file_path), exist_ok=True)\n",
    "            output_handles[tag_file_cell_type_name] = open(tag_file_path, \"w\")\n",
    "\n",
    "\n",
    "    print(\"open local_fragment_file {}\".format(local_fragment_file))\n",
    "    with gzip.open(local_fragment_file, \"rt\") as infile:\n",
    "        missing_bc = 0\n",
    "        num_of_lines_written=0\n",
    "        for line_number, line in enumerate(infile, start=1):\n",
    "            # # debug\n",
    "            # if line_number > 10:\n",
    "            #     continue \n",
    "\n",
    "            out_list = split_fragment_line_string(line)\n",
    "            bc = out_list[-1]\n",
    "            # print(\"XXXXXX bc is {}\".format(bc))\n",
    "#             Austin output: chrom, start, end, bc, rem = line.rstrip('\\n').split('\\t', 5)\n",
    "            out_line_to_print = f\"{out_list[0]}\\t{out_list[1]}\\t{out_list[2]}\\t{bc}\\t{out_list[-2]}\\n\"\n",
    "            # print(\"XXXXXX out_line_to_print is {}\".format(out_line_to_print))\n",
    "#             chr1\t10007\t10175\tENCSR023FME#ENCSR023FME_GAAGGTTCAAAGTGTCAGTCAA\t1\n",
    "            num_of_lines_written +=1\n",
    "            returnTagAlign = convert_fragment_line_to_tagAlign(out_line_to_print)\n",
    "            # print(\"XXXXXX returnTagAlign is {}\".format(returnTagAlign))\n",
    "            # write to the relevant cell type file\n",
    "            bc_exists_in_cell_type_atac_dataset = df_cell_types_for_atac_dataset[df_cell_types_for_atac_dataset['cell_id']==bc]\n",
    "            # print(\"bc_exists_in_cell_type_atac_dataset is {}\".format(bc_exists_in_cell_type_atac_dataset))\n",
    "            # print(\"len(bc_exists_in_cell_type_atac_dataset) is {} for bc {}\".format(len(bc_exists_in_cell_type_atac_dataset),bc))\n",
    "            if len(bc_exists_in_cell_type_atac_dataset) ==1:\n",
    "                tag_file_cell_type_id = df_cell_types_for_atac_dataset.loc[df_cell_types_for_atac_dataset['cell_id'] == bc, 'cell_type_id'].iloc[0]\n",
    "                # print(\"tag_file_cell_type_id is {}\".format(tag_file_cell_type_id))\n",
    "                output_handles[tag_file_cell_type_id].write(returnTagAlign)\n",
    "            else:\n",
    "                missing_bc+=1\n",
    "\n",
    "        # if the file is not closed. data will be missing\n",
    "        for tag_file_path, tag_file_cell_type_name in output_tagAlign_files_with_names:\n",
    "            # print(\"tag_file_path is {}\".format(tag_file_path))\n",
    "            output_handles[tag_file_cell_type_name].close()            \n",
    "        print(\"finished clustering local_fragment_file {} by cell type. for types {}\".format(local_fragment_file, cell_type_id_names_for_atac_dataset))\n",
    "        print(\"total missing bc are {}\".format(missing_bc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea6f9d5b-acef-458d-b5c7-aa8186821b7d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
